src/main/java/dev/su5ed/mffs/blockentity/CoercionDeriverBlockEntity.java

```diff
@@ -86,52 +84,91 @@ protected void addModuleSlots(List<? super InventorySlot> list) {
-                    }
-                    this.processTime = Math.max(--this.processTime, 0);
+                dischargeItemIntoSelf(this.batterySlot.getItem()); //TODO shouldn't this be disabled with enableElectricity flag
+
+                if (this.energy.canExtract(MFFSConfig.COMMON.coercionDriverFePerFortron.get()) || !MFFSConfig.COMMON.enableElectricity.get() && hasFuel()) {
+                    produceFortron();
+                    consumeFuel();
                 }
             }
         }
     }
 
+    private void consumeFuel() {
+        // TODO Fuel display
+        if (this.processTime == 0 && hasFuel()) {
+            this.fuelSlot.getItem().shrink(1);
+            this.processTime = MFFSConfig.COMMON.catalystBurnTime.get() * Math.max(getModuleCount(ModModules.SCALE) / 20, 1); //TODO why 20
+        }
+        this.processTime = Math.max(--this.processTime, 0);
+    }
+
+    private void produceFortron() {
+        final int fortronOutput = calculateFortronProduction();
+        final int fortronStored = fortronProducedLastTick = this.fortronStorage.insertFortron(fortronOutput, false);
+        final int asEnergy = fortronStored * MFFSConfig.COMMON.coercionDriverFePerFortron.get();
+        this.energy.extractEnergy(asEnergy, false);
+    }
+
+    /**
+     * Predicted fortron to be produced next energy tick
+     *
+     * @return fortron(ml)
+     */
+    public int calculateFortronProduction() {
+        final int spaceLeft = this.fortronStorage.getFortronCapacity() - this.fortronStorage.getStoredFortron();
+        final int maxFortronFromEnergy = this.energy.getEnergyStored() / MFFSConfig.COMMON.coercionDriverFePerFortron.get();
+        return Math.min(maxFortronFromEnergy, Math.min(getMaxFortronProducedPerTick(), spaceLeft));
+    }
+
+    private void convertFortronIntoEnergy() {
+        final int energyPerFortron = MFFSConfig.COMMON.coercionDriverFePerFortron.get() - MFFSConfig.COMMON.coercionDriverFortronToFeLoss.get();
+
+        // Only run if we can withdraw at least 1 fortron
+        if (this.energy.getEnergyStored() + energyPerFortron < this.energy.getMaxEnergyStored()) {
+            //Calculate upper limits per tick
+            final int maxFortronOut = this.fortronStorage.extractFortron(getMaxFortronProducedPerTick(), true);
+            final int maxEnergyOut = maxFortronOut * energyPerFortron;
+
+            //Calculate amount of energy that can actually be moved
+            final int maxEnergyReceived = this.energy.receiveEnergy(maxEnergyOut, true);
+
+            // Calculate actual values to move, round down to avoid material loss
+            final int fortronToRemove = (int)Math.floor(maxEnergyReceived / (float)energyPerFortron);
+
+            // Apply values
+            this.energy.receiveEnergy(this.fortronStorage.extractFortron(fortronToRemove, true) * energyPerFortron, true);
```

Comment:
Found a bug here that seems to be caused by using simulate=true despite trying to actually extract energy.